okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/zip.kt

```diff
@@ -175,22 +175,60 @@ fun open(zipPath: Path, fileSystem: FileSystem): ZipFileSystem {
   // public API doesn't allow us to throw IOException except from the constructor
   // or from getInputStream.
   cursor.seek(record.centralDirOffset)
-  val entries = mutableMapOf<Path, ZipEntry>()
+  val entries = mutableListOf<ZipEntry>()
   for (i in 0 until record.numEntries) {
     val newEntry = source.readEntry(
       isZip64 = zip64EocdRecordOffset != -1L
     )
     if (newEntry.localHeaderRelOffset >= record.centralDirOffset) {
       throw IOException("Local file header offset is after central directory")
     }
-    val entryName = newEntry.name
-    val path = "/".toPath().div(entryName)
-    if (entries.put(path, newEntry) != null) {
-      throw IOException("Duplicate path: $entryName")
+    entries += newEntry
+  }
+
+  val index = buildIndex(entries)
+
+  return ZipFileSystem(zipPath, fileSystem, index, comment)
+}
+
+/**
+ * Returns a map containing all of [entries], plus parent entries required so that all entries
+ * (other than the file system root `/`) have a parent.
+ */
+@ExperimentalFileSystem
+private fun buildIndex(entries: List<ZipEntry>): Map<Path, ZipEntry> {
+  val result = mutableMapOf<Path, ZipEntry>()
+
+  // Iterate in sorted order so each path is preceded by its parent.
+  for (entry in entries.sortedBy { it.canonicalPath }) {
+    if (result.put(entry.canonicalPath, entry) != null) {
+      throw IOException("Duplicate path: ${entry.canonicalPath}")
```

Comment:
There is a suggestion the base format allows this but java makes it fail

https://stackoverflow.com/a/3113584/1542667

Reply:
Agreed. It’s useful validation to reject it.

Reply:
Postels law seems relevent when consumung externally produced files?

Reply:
Not in this case. The detection is preventative!

https://androidstudygroup.slack.com/archives/D01M90K55LK/p1613431367008700

Reply:
Not in this case. The detection is preventative!

https://nakedsecurity.sophos.com/2013/07/10/anatomy-of-a-security-hole-googles-android-master-key-debacle-explained/